### ISSUE TITLE
ppptp filter problem

### ISSUE THREAD

**Issue by Scorpikor** [fail2ban/fail2ban#2446]
Hi all!
I have log in /var/log/syslog :
```
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Client 46.161.27.42 control connection started
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Starting call (launching pppd, opening GRE)
Jun 14 14:28:33 cyclon pppd[26669]: Plugin /usr/lib/pptpd/pptpd-logwtmp.so loaded.
Jun 14 14:28:33 cyclon pppd[26669]: pppd 2.4.7 started by root, uid 0
Jun 14 14:28:33 cyclon pppd[26669]: Using interface ppp1
Jun 14 14:28:33 cyclon pppd[26669]: Connect: ppp1 <--> /dev/pts/3
Jun 14 14:28:33 cyclon pptpd[26666]: GRE: Bad checksum from pppd.
Jun 14 14:28:33 cyclon pppd[26669]: appear to have received our own echo-reply!
Jun 14 14:28:33 cyclon pppd[26669]: No CHAP secret found for authenticating Admin
Jun 14 14:28:33 cyclon pppd[26669]: Peer Admin failed CHAP authentication
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: EOF or bad error reading ctrl packet length.
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: couldn't read packet header (exit)
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: CTRL read failed
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Reaping child PPP[26669]
Jun 14 14:28:33 cyclon pppd[26669]: Modem hangup
Jun 14 14:28:33 cyclon pppd[26669]: Connection terminated.
Jun 14 14:28:33 cyclon pppd[26669]: Exit.
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Client 46.161.27.42 control connection finished

```
I try to write filter:
/etc/fail2ban/filter.d/pptpd.local
```
[Definition]
# to get ID 26669 from:
# Jun 14 14:28:33 cyclon pppd[26669]: Peer Admin failed CHAP authentication
pref1 = (?<=pppd\[).+?(?=\]: Peer.+?failed CHAP authentication)

# to get ID 26666 from:
# Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Reaping child PPP[26669]
pref2 = (?<=pptpd\[).+?(?=\]: CTRL: Reaping child PPP\[%(pref1)s\])

# to get IP from:
# Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Client 46.161.27.42 control connection finished
failregex = pptpd\[%(pref2)s\]: CTRL: Client <HOST> control connection finished

```

Is it right way or I can`t do like this ? What is the right way to run this filter?
Of course this filter do not work :)

**Comment by sebres** [comment]
People, it is really enough!
In our issue template we are asking fail2ban version not for nothing.
### It is important, so provide it!

**Comment by Scorpikor** [comment]
Sorry, 
fail2ban-server -V
Fail2Ban v0.10.2

**Comment by sebres** [comment]
Since fail2ban 0.10 allow multi-line parsing (*MLFID*-related parsing using tag `<F-MLFID>` instead of buffering with maxlines), this can be used to identify lines belonging to single session (to find IP if the failure message does not contain it).

 Here is your filter:
```ini
[Definition]

prefregex = ^\s*\S+\s*ppt?pd\[<F-MLFID>\d+</F-MLFID>\]: <F-CONTENT>.+</F-CONTENT>$

failregex = ^Peer <F-USER>\S+|.+?</F-USER> failed CHAP authentication$
            ^CTRL: Client <ADDR> control connection <F-NOFAIL>started</F-NOFAIL>$
            ^CTRL: Client <ADDR> control connection <F-NOFAIL><F-MLFFORGET>finished</F-MLFFORGET></F-NOFAIL>$
```

The `prefregex` is used to match prefix (common part of all messages) and common session `MLFID`.
First regex in `failregex` matches a failure (but because there is no IP, there are...)
2nd and 3rd regex's in `failregex` are helper to capture IP only (don't cause a failure).

Also note `ppt?pd` will match `pptpd` as well as `pppd`,

If a line produced in success case is known too, and #2279 will be released (and you'd get update to 0.10.5), this filter can be additionally extended to match this message using tag `<F-MLFGAINED>`, with something like `...<F-MLFGAINED>connected</F-MLFGAINED>...` in order to signal a connect is succeeded.

**Comment by Aqualion509** [comment]
Skorpikor, tell please, does this build work for you? 
I have similar problem with pppd\pptpd and this solution doesn't work :(
My fail2ban-server -V is: 0.11.2
I think problem with different id's :
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Client 46.161.27.42 control connection started
Jun 14 14:28:33 cyclon pppd[26669]: Peer Admin failed CHAP authentication
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Reaping child PPP[26669]
Jun 14 14:28:33 cyclon pptpd[26666]: CTRL: Client 46.161.27.42 control connection finished

**Comment by sebres** [comment]
> I think problem with different id's

Yes. 
Value `pptpd[26666]` (for what the IP got stored with NOI-FAIL helpers) and `pppd[26669]` (introducing the failure) are different.
And unfortunately there is no possibility to link the IDs or to obtain/replace it from messages like `pptpd[26666]: CTRL: Reaping child PPP[26669]` at the moment.

If authentication is a mandatory process, and it takes place almost immediately (so the messages following in small distance), you'd try to combine this `MLFID` procedure with multiline window (`maxlines = 10` and RE without `prefregex` searching for both messages at once `failed ... authentication` and ` Reaping child PPP\[(?P=mlfid)\]` and with `<SKIPLINES>` between both).
But it is bad practice (such filter were too error prone).

Another (better) variant would be to find a message (with same MLFID as in connection started/finished) occurring only if authentication failed (to avoid false positives for legitimate attempts).
So in this example `CTRL read failed` used as such a message:
```ini
[Definition]

prefregex = ^\s*\S+\s*ppt?pd\[<F-MLFID>\d+</F-MLFID>\]: <F-CONTENT>.+</F-CONTENT>$

failregex = ^CTRL: CTRL read failed
            ^CTRL: Client <ADDR> control connection <F-NOFAIL>started</F-NOFAIL>$
            ^CTRL: Client <ADDR> control connection <F-NOFAIL><F-MLFFORGET>finished</F-MLFFORGET></F-NOFAIL>$
```

**Comment by Aqualion509** [comment]
Sebres, thanks for your answer.
Solution above, with string "CTRL: CTRL read failed" is suitable for us and working good.
Good Luck!

**Comment by BilalBhatti** [comment]
> ```ini
> prefregex = ^\s*\S+\s*ppt?pd\[<F-MLFID>\d+</F-MLFID>\]: <F-CONTENT>.+</F-CONTENT>$
> ```

I see the real issue here. Actually session MLFID increments to +1 (or sometimes +2) as a child session. This causes script not to find the related IP of messages using failregex. The only link between child and parent session is CTRL: Reaping child PPP which we could use somehow. Below are the logs of failure:

```
pptpd[**14178**]: CTRL: Client 43.24.8.6 control connection **started**
pptpd[14178]: CTRL: Starting call (launching pppd, opening GRE)
pppd[14179]: Plugin /usr/lib/pptpd/pptpd-logwtmp.so loaded.
pppd[14179]: pppd 2.4.7 started by root, uid 0
pppd[14179]: Using interface ppp0
pppd[14179]: Connect: ppp0 <--> /dev/pts/4
pppd[**14179**]: Peer test1 **failed** CHAP authentication
pppd[14179]: Connection terminated.
pppd[14179]: Exit.
pptpd[**14178**]: CTRL: Reaping child PPP[**14179**]
pptpd[**14178**]: CTRL: Client 43.245.8.64 control connection **finished**
```

For successful connection:

```
pptpd[**13907**]: CTRL: Client 43.24.8.6 control connection **started**
pptpd[13907]: CTRL: Starting call (launching pppd, opening GRE)
pppd[13908]: Plugin /usr/lib/pptpd/pptpd-logwtmp.so loaded.
pppd[13908]: pppd 2.4.7 started by root, uid 0
pppd[13908]: Using interface ppp0
pppd[13908]: Connect: ppp0 <--> /dev/pts/4
pppd[**13908**]: peer from calling number 43.24.8.6 **authorized**
pppd[13908]: MPPE 128-bit stateless compression enabled
pppd[13908]: LCP terminated by peer (MPPE disabled)
pppd[13908]: Connect time 0.3 minutes.
pppd[13908]: Sent 1769 bytes, received 3072 bytes.
pptpd[13907]: CTRL: Reaping child PPP[13908]
pppd[13908]: Modem hangup
pppd[13908]: Connection terminated.
pppd[13908]: Exit.
pptpd[**13907**]: CTRL: Client 43.24.8.6 control connection **finished**
```

pptpd[Parent-ID]: Reaping child PPP[Child-ID]

Parent-ID here should be used to find IP and Child-ID should be used to check if authentication passed or failed. This will make it fool proof against all or max false positives as well.

**Comment by sebres** [comment]
> I see the real issue here. Actually session MLFID increments to +1 (or sometimes +2) as a child session.

Yes this is the real issue. And as already said in https://github.com/fail2ban/fail2ban/issues/2446#issuecomment-1036504810:
*unfortunately there is no possibility to link the IDs or to obtain/replace it from messages like `pptpd[26666]: CTRL: Reaping child PPP[26669]` at the moment.*
The second solution I suggested is a "workaround" rather. But it seems to work in certain cases.

You wrote in #2754:
> It still bans IP after successful connection

But neither with a filter example from there nor with example above it would find a failure at all.
The RE matching `connection started|finished` are masked by `<F-NOFAIL>` wich means that this is not a failure (it is the helper to obtain IP address and link it with a MLFID).
So only RE with `CTRL read failed` would match a failure and your log excerpts seem not to have such messages neither for failed nor for succeeded connection.
So I'd understand if you'd say it doesn't find failures and doesn't ban at all. But not why it must ban something or even have false positives.

Considering current restriction of fail2ban (impossibility to link different MLFID), if one'd filter your log excerpts by MLFID (which messages containing IPs), we'll see this:
```
# failed:
pptpd[14178]: CTRL: Client 192.0.2.6 control connection started
pptpd[14178]: CTRL: Starting call (launching pppd, opening GRE)
pptpd[14178]: CTRL: Reaping child PPP[14179]
pptpd[14178]: CTRL: Client 192.0.2.6 control connection finished

# succeeded:
pptpd[13907]: CTRL: Client 192.0.2.6 control connection started
pptpd[13907]: CTRL: Starting call (launching pppd, opening GRE)
pptpd[13907]: CTRL: Reaping child PPP[13908]
pptpd[13907]: CTRL: Client 192.0.2.6 control connection finished
```
what shows us the same picture, so it would be impossible to use this alone.

Usage of multiline filter with larger parse window (buffering lines `maxlines` > 1) is also hardly suitable here, because too error-prone since an intruder need to make a small pause or do few connections simultaneously to avoid match.

Although message `pppd[13908]: peer from calling number 192.0.2.6 authorized` contain the IP, so one could theoretically write a different filter using IP address as a MLFID and `<F-MLFGAINED>` introduced in #2279 (released with 0.10.5 / 0.11.1).

```ini
[Definition]

prefregex = ^\s*(?:\S+ )?ppt?pd\[[^\]]+\]: <F-CONTENT>.+</F-CONTENT>$

failregex = ^CTRL: Client <F-MLFID>\S+</F-MLFID> control connection started
            ^CTRL: Client <F-MLFID><ADDR></F-MLFID> control connection <F-NOFAIL><F-MLFFORGET>finished</F-MLFFORGET></F-NOFAIL>
            ^peer from calling number <F-MLFID><ADDR></F-MLFID> <F-MLFGAINED>authorized</F-MLFGAINED>
```
But! The problem here is that fail2ban had a small inconsistency by handling of `<F-MLFGAINED>` state (it did not reset pending failures directly by processing of that RE), so unfortunately also this filter wouldn't work correctly at the moment.
This is fixed now in e2d50f38a6ef2511fee6b49f42b98f6d867625b2, but...

Just I don't like such filters at all for many reasons (which I won't publish here to prevent insider details may reach some script-kiddies). But main reason is that a legitimate user (e. g. "guest") which is basically allowed to authorize could theoretically exercise a bruteforce attack for another accounts (e. g. "root") and after X attempts make a successful authorization for guest account in order to produce `peer ... authorized` and so would force fail2ban to "forget" the failed attempts by connection finished.

There are only 2 possibility to process such log (in absence of `CTRL read failed` message):
1. either this pppd need to be extended for example provide the IP by authentication failures, e. g. something like this:
```diff
-pppd[14179]: Peer test1 failed CHAP authentication
+pppd[14179]: Peer 192.0.2.6 failed CHAP authentication
```
2. or fail2ban needs to be extended to support MLFID mapping (e. g. `14178` and `14179` became synonyms after `pptpd[14178]: CTRL: Reaping child PPP[14179]`).

Sorry, but I don't see at the moment any reason why option 2 must be preferable than option 1. 
Moreover I am personally and very strongly convinced that extending of pppd (or even pptpd) with IP by failure is much better option (especially because it'd makes the things simpler and because the IP seems to be basically available there, since `authorized` message does contain it).

**Comment by BilalBhatti** [comment]
Thankyou @sebres 

Finally its working fine (actually observed zero failures or zero false positives) so far. Cloned latest version from this repo (since you told that small inconsistency by handling of <F-MLFGAINED> has been fixed and used below code to finally ban unwanted PPTP sessions. So far working flawlessly.

```
[Definition]

prefregex = ^\s*(?:\S+ )?ppt?pd\[[^\]]+\]: <F-CONTENT>.+</F-CONTENT>$

failregex = ^CTRL: Client <F-MLFID>\S+</F-MLFID> control connection started
            ^CTRL: Client <F-MLFID><ADDR></F-MLFID> control connection <F-NOFAIL><F-MLFFORGET>finished</F-MLFFORGET></F-NOFAIL>
            ^peer from calling number <F-MLFID><ADDR></F-MLFID> <F-MLFGAINED>authorized</F-MLFGAINED>
```

For instance, it successfully banned below bot:

```
pptpd[4691]: CTRL: Client 91.191.209.235 control connection started
pptpd[4691]: CTRL: Starting call (launching pppd, opening GRE)
pppd[4693]: Plugin /usr/lib/pptpd/pptpd-logwtmp.so loaded.
pppd[4693]: pppd 2.4.7 started by root, uid 0
pppd[4693]: Using interface ppp0
pppd[4693]: Connect: ppp0 <--> /dev/pts/3
pppd[4693]: appear to have received our own echo-reply!
pppd[4693]: No CHAP secret found for authenticating test
pppd[4693]: Peer test failed CHAP authentication
pptpd[4691]: CTRL: EOF or bad error reading ctrl packet length.
pptpd[4691]: CTRL: couldn't read packet header (exit)
pptpd[4691]: CTRL: CTRL read failed
pptpd[4691]: CTRL: Reaping child PPP[4693]
pppd[4693]: Modem hangup
pppd[4693]: Connection terminated.
pppd[4693]: Exit.
pptpd[4691]: CTRL: Client 91.191.209.235 control connection finished
```

And ignored following:

```
pptpd[2940]: CTRL: Client 59.10.80.15 control connection started
pptpd[2940]: CTRL: Starting call (launching pppd, opening GRE)
pppd[2941]: Plugin /usr/lib/pptpd/pptpd-logwtmp.so loaded.
pppd[2941]: pppd 2.4.7 started by root, uid 0
pppd[2941]: Using interface ppp0
pppd[2941]: Connect: ppp0 <--> /dev/pts/3
pppd[2941]: peer from calling number 59.100.80.150 authorized
pppd[2941]: MPPE 128-bit stateless compression enabled
pppd[2941]: Cannot determine ethernet address for proxy ARP
pppd[2941]: local  IP address 192.168.10.120
pppd[2941]: remote IP address 192.168.20.1
pppd[2941]: LCP terminated by peer (MPPE disabled)
pppd[2941]: Connect time 0.1 minutes.
pppd[2941]: Sent 6676 bytes, received 7847 bytes.
pptpd[2940]: CTRL: EOF or bad error reading ctrl packet length.
pptpd[2940]: CTRL: couldn't read packet header (exit)
pptpd[2940]: CTRL: CTRL read failed
pptpd[2940]: CTRL: Reaping child PPP[2941]
pppd[2941]: Modem hangup
pppd[2941]: Connection terminated.
pppd[2941]: Exit.
pptpd[2940]: CTRL: Client 59.10.80.15 control connection finished
```

Thankyou again!
